sql-plugin/src/main/scala/com/nvidia/spark/rapids/window/GpuCachedDoublePassWindowExec.scala

```diff
@@ -0,0 +1,384 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.window
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+import ai.rapids.cudf
+import ai.rapids.cudf.{NvtxColor, Scalar}
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.Arm.{closeOnExcept, withResource, withResourceIfAllowed}
+import com.nvidia.spark.rapids.RmmRapidsRetryIterator.{splitSpillableInHalfByRows, withRetry}
+import com.nvidia.spark.rapids.ScalableTaskCompletion.onTaskCompletion
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.expressions.{Expression, NamedExpression, SortOrder}
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.rapids.aggregate.GpuAggregateExpression
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+
+class FixerPair(op: GpuUnboundToUnboundWindowWithFixer) extends AutoCloseable {
+  var fixing: BatchedUnboundedToUnboundedWindowFixer = op.newUnboundedToUnboundedFixer
+  var collecting: BatchedUnboundedToUnboundedWindowFixer = op.newUnboundedToUnboundedFixer
+
+  def updateState(scalar: Scalar): Unit = {
+    collecting.updateState(scalar)
+  }
+
+  def fixUp(samePartitionMask: Either[cudf.ColumnVector, Boolean],
+      column: cudf.ColumnVector): cudf.ColumnVector =
+    fixing.fixUp(samePartitionMask, column)
+
+  def swap(): Unit = {
+    val tmp = fixing
+    tmp.reset()
+    fixing = collecting
+    collecting = tmp
+  }
+
+  override def close(): Unit = {
+    fixing.close()
+    collecting.close()
+  }
+}
+
+/**
+ * An iterator that can do aggregations on window queries that need a small amount of
+ * information from all of the batches to update the result in a second pass. It does this by
+ * having the aggregations be instances of GpuUnboundToUnboundWindowWithFixer
+ * which can fix up the window output for unbounded to unbounded windows.
+ * Because of this there is no requirement about how the input data is batched, but it  must
+ * be sorted by both partitioning and ordering.
+ */
+class GpuCachedDoublePassWindowIterator(
```

Comment:
Should we move `GpuCachedDoublePassWindowIterator` into its own file?  Just as `GpuBatchedWindowIterator` has its own source file.

Reply:
Ah, maybe not, since `GpuCachedDoublePassWindowExec` is tiny on its own.

Reply:
We could, but I thought it was tied directly to `GpuCachedDoublePassExec`, just like `GpuBatchedBoundedWindowIterator` is in the same file as `GpuBatchedBoundedWindowExec`. `GpuBatchedWindowIterator` probably should be renamed because it is really a Utils object class and not an actual Iterator.